src/functions.php

```diff
@@ -393,14 +393,40 @@ function extract_session_from_options(array $options)
 }
 
 /**
- * Performs server selection, respecting the server a session may be pinned to
+ * Returns the readPreference option if it is set and valid.
+ *
+ * @internal
+ * @param array $options
+ * @return ReadPreference|null
+ */
+function extract_read_preference_from_options(array $options)
+{
+    if (! isset($options['readPreference']) || ! $options['readPreference'] instanceof ReadPreference) {
+        return null;
+    }
+
+    return $options['readPreference'];
+}
+
+/**
+ * Performs server selection, respecting the readPreference and session options
+ * (if given)
  *
  * @internal
  * @return Server
  */
-function select_server(Manager $manager, ReadPreference $readPreference = null, Session $session = null)
+function select_server(Manager $manager, array $options)
 {
-    $server = $session !== null ? $session->getServer() : null;
+    $session = extract_session_from_options($options);
+    if ($session !== null && $session->getServer() !== null) {
+        return $session->getServer();
+    }
+
+    $readPreference = extract_read_preference_from_options($options);
+    if (! $readPreference instanceof ReadPreference) {
```

Comment:
It makes no technical difference, but this seems inconsistent with the previous `if` condition. Would it make sense to use `$readPreference === null` here?

Reply:
Ha...I thought I'd be smart and be type safe here since we don't have return types yet. If you don't mind I'd rather change the call above to also check for `$session` being an `MongoDB\Driver\Session` instance instead of just checking for it not being `null`.

Reply:
I'm totally on board with `instanceof` over `!== null`. Go for it.